gossipsub_test.go

```diff
@@ -0,0 +1,857 @@
+
+		psubs[owner].Publish(topic, msg)
+
+		for _, sub := range msgs[i] {
+			got, err := sub.Next(ctx)
+			if err != nil {
+				t.Fatal(sub.err)
+			}
+			if !bytes.Equal(msg, got.Data) {
+				t.Fatal("got wrong message!")
+			}
+		}
+	}
+}
+
+func TestMixedGossipsub(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	hosts := getNetHosts(t, ctx, 30)
+
+	gsubs := getGossipsubs(ctx, hosts[:20])
+	fsubs := getPubsubs(ctx, hosts[20:])
+	psubs := append(gsubs, fsubs...)
+
+	var msgs []*Subscription
+	for _, ps := range psubs {
+		subch, err := ps.Subscribe("foobar")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		msgs = append(msgs, subch)
+	}
+
+	sparseConnect(t, hosts)
+
+	// wait for heartbeats to build mesh
+	time.Sleep(time.Second * 2)
+
+	for i := 0; i < 100; i++ {
+		msg := []byte(fmt.Sprintf("%d it's not a floooooood %d", i, i))
+
+		owner := rand.Intn(len(psubs))
+
+		psubs[owner].Publish("foobar", msg)
+
+		for _, sub := range msgs {
+			got, err := sub.Next(ctx)
+			if err != nil {
+				t.Fatal(sub.err)
+			}
+			if !bytes.Equal(msg, got.Data) {
+				t.Fatal("got wrong message!")
+			}
+		}
+	}
+}
+
+func TestGossipsubMultihops(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	hosts := getNetHosts(t, ctx, 6)
+
+	psubs := getGossipsubs(ctx, hosts)
+
+	connect(t, hosts[0], hosts[1])
+	connect(t, hosts[1], hosts[2])
+	connect(t, hosts[2], hosts[3])
+	connect(t, hosts[3], hosts[4])
+	connect(t, hosts[4], hosts[5])
+
+	var subs []*Subscription
+	for i := 1; i < 6; i++ {
+		ch, err := psubs[i].Subscribe("foobar")
+		if err != nil {
+			t.Fatal(err)
+		}
+		subs = append(subs, ch)
+	}
+
+	// wait for heartbeats to build mesh
+	time.Sleep(time.Second * 2)
+
+	msg := []byte("i like cats")
+	err := psubs[0].Publish("foobar", msg)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// last node in the chain should get the message
+	select {
+	case out := <-subs[4].ch:
+		if !bytes.Equal(out.GetData(), msg) {
+			t.Fatal("got wrong data")
+		}
+	case <-time.After(time.Second * 5):
+		t.Fatal("timed out waiting for message")
+	}
+}
+
+func TestGossipsubTreeTopology(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	hosts := getNetHosts(t, ctx, 10)
+	psubs := getGossipsubs(ctx, hosts)
+
+	connect(t, hosts[0], hosts[1])
+	connect(t, hosts[1], hosts[2])
+	connect(t, hosts[1], hosts[4])
+	connect(t, hosts[2], hosts[3])
+	connect(t, hosts[0], hosts[5])
+	connect(t, hosts[5], hosts[6])
+	connect(t, hosts[5], hosts[8])
+	connect(t, hosts[6], hosts[7])
+	connect(t, hosts[8], hosts[9])
+
+	/*
+		[0] -> [1] -> [2] -> [3]
+		 |      L->[4]
+		 v
+		[5] -> [6] -> [7]
+		 |
+		 v
+		[8] -> [9]
+	*/
+
+	var chs []*Subscription
+	for _, ps := range psubs {
+		ch, err := ps.Subscribe("fizzbuzz")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		chs = append(chs, ch)
+	}
+
+	// wait for heartbeats to build mesh
+	time.Sleep(time.Second * 2)
```

Comment:
Why do we always have to wait for heartbeats? Is it because we don't send out subscription notices immediately now?

Reply:
Well, subscription notices are still sent immediately, just retried if they fail.

Now that I think of it, we can probably greatly reduce this delay in almost all the tests -- maybe down to say 100ms, just enough for announcements to go out. 

My rationale for this delay was to wait a couple of heartbeats to avoid interference from nodes who have joined but haven't seen any peer announcements yet. Also, I wanted to avoid interference from the overlay construction, but it should still be connected after the announcements get sent and nodes pick their peers.

Reply:
Actually, there is a genuine concern that is an artifact of the concurrent Join from subscriptions.
If we subscribe all the nodes together, then they won't have any peer announcements when the do the Join, and they'll have to wait a heartbeat before they start adding peers to the mesh.
We can avoid this if we add a small (say 10ms) delay after each subscription.

Reply:
Hrm... I'm very skeptical of 'fixing' things by adding delays.

Reply:
Also, in many of the tests the subsscriptions are created before connecting the network, which means that all nodes start empty and build the mesh purely in the heartbeat

Reply:
@vyzo It's really really bad practice to use delays like this, especially without a select statement to escape out of it if the context is canceled.

Reply:
@paralin , this is just a test that _needs_ a delay -- and there is nothing to cancel the context so a select would be totally useless.

Reply:
@vyzo Gotcha, I notice now it's a test.